tests/functional/oauth_webchannel_keys.js

```diff
@@ -0,0 +1,510 @@
+  var OLD_ENOUGH_YEAR = TOO_YOUNG_YEAR - 1;
+  var user;
+  var email;
+  var client;
+  var ANIMATION_DELAY_MS = 1000;
+
+  var listenForSyncCommands = FxDesktopHelpers.listenForFxaCommands;
+  var testIsBrowserNotifiedOfSyncLogin = FxDesktopHelpers.testIsBrowserNotifiedOfLogin;
+
+
+  /**
+   * This suite tests the WebChannel functionality for delivering encryption keys
+   * in the OAuth signin and signup cases. It uses a CustomEvent "WebChannelMessageToChrome"
+   * to finish OAuth flows
+   */
+
+  function testIsBrowserNotifiedOfLogin(context, options) {
+    options = options || {};
+    return FunctionalHelpers.testIsBrowserNotified(context, 'oauth_complete', function (data) {
+      assert.ok(data.redirect);
+      assert.ok(data.code);
+      assert.ok(data.state);
+      // All of these flows should produce encryption keys.
+      assert.ok(data.keys);
+      assert.equal(data.closeWindow, options.shouldCloseTab);
+    });
+  }
+
+  function openFxaFromRp(context, page, additionalQueryParams) {
+    var queryParams = '&webChannelId=test&keys=true';
+    for (var key in additionalQueryParams) {
+      queryParams += ('&' + key + '=' + additionalQueryParams[key]);
+    }
+    return FunctionalHelpers.openFxaFromRp(context, page, queryParams);
+  }
+
+
+  registerSuite({
+    name: 'oauth web channel keys',
+
+    beforeEach: function () {
+      email = TestHelpers.createEmail();
+      user = TestHelpers.emailToUser(email);
+
+      client = new FxaClient(AUTH_SERVER_ROOT, {
+        xhr: nodeXMLHttpRequest.XMLHttpRequest
+      });
+
+      return FunctionalHelpers.clearBrowserState(this, {
+        contentServer: true,
+        '123done': true
+      });
+    },
+
+    'signup, verify same browser with original tab open': function () {
+      var self = this;
+
+      return openFxaFromRp(self, 'signup')
+        .execute(FunctionalHelpers.listenForWebChannelMessage)
+
+        .then(function () {
+          return FunctionalHelpers.fillOutSignUp(self, email, PASSWORD, OLD_ENOUGH_YEAR);
+        })
+
+        .findByCssSelector('#fxa-confirm-header')
+        .end()
+
+        .then(function () {
+          return FunctionalHelpers.openVerificationLinkSameBrowser(
+                      self, email, 0);
+        })
+        .switchToWindow('newwindow')
+        // wait for the verified window in the new tab
+        .findById('fxa-sign-up-complete-header')
+        // the verification tab will almost always be the one to complete the flow.
+        // XXX TODO: can we not fail the test if the other tab wins?
+        // XXX TODO: or, ensure that the original tab always wins if it's open?
+        .then(testIsBrowserNotifiedOfLogin(self, { shouldCloseTab: false }))
+        .end()
+
+        .closeCurrentWindow()
+        // switch to the original window
+        .switchToWindow('')
+
+        .findById('fxa-sign-up-complete-header')
+        .end();
+    },
+
+    'signup, verify same browser with original tab closed': function () {
```

Comment:
nit: naming, `with` vs adding a `,` related to https://github.com/mozilla/fxa-content-server/pull/2385/files#r30045417